---
author-meta:
- Jineta Banerjee
- Jaclyn N Taroni
- Robert J Allaway
- Deepashree Venkatesh Prasad
- Justin Guinney
- Casey Greene
bibliography:
- content/manual-references.json
date-meta: '2023-03-01'
header-includes: "<!--\nManubot generated metadata rendered from header-includes-template.html.\nSuggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html\n-->\n<meta name=\"dc.format\" content=\"text/html\" />\n<meta name=\"dc.title\" content=\"Machine learning in rare disease\" />\n<meta name=\"citation_title\" content=\"Machine learning in rare disease\" />\n<meta property=\"og:title\" content=\"Machine learning in rare disease\" />\n<meta property=\"twitter:title\" content=\"Machine learning in rare disease\" />\n<meta name=\"dc.date\" content=\"2023-03-01\" />\n<meta name=\"citation_publication_date\" content=\"2023-03-01\" />\n<meta name=\"dc.language\" content=\"en-US\" />\n<meta name=\"citation_language\" content=\"en-US\" />\n<meta name=\"dc.relation.ispartof\" content=\"Manubot\" />\n<meta name=\"dc.publisher\" content=\"Manubot\" />\n<meta name=\"citation_journal_title\" content=\"Manubot\" />\n<meta name=\"citation_technical_report_institution\" content=\"Manubot\" />\n<meta name=\"citation_author\" content=\"Jineta Banerjee\" />\n<meta name=\"citation_author_institution\" content=\"Sage Bionetworks\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-1775-3645\" />\n<meta name=\"citation_author\" content=\"Jaclyn N Taroni\" />\n<meta name=\"citation_author_institution\" content=\"Childhood Cancer Data Lab, Alex\u2019s Lemonade Stand Foundation\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-4734-4508\" />\n<meta name=\"citation_author\" content=\"Robert J Allaway\" />\n<meta name=\"citation_author_institution\" content=\"Sage Bionetworks\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-3573-3565\" />\n<meta name=\"twitter:creator\" content=\"@allawayr\" />\n<meta name=\"citation_author\" content=\"Deepashree Venkatesh Prasad\" />\n<meta name=\"citation_author_institution\" content=\"Childhood Cancer Data Lab, Alex\u2019s Lemonade Stand Foundation\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-5756-4083\" />\n<meta name=\"citation_author\" content=\"Justin Guinney\" />\n<meta name=\"citation_author_institution\" content=\"Sage Bionetworks\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-1477-1888\" />\n<meta name=\"citation_author\" content=\"Casey Greene\" />\n<meta name=\"citation_author_institution\" content=\"Department of Systems Pharmacology and Translational Therapeutics, Perelman School of Medicine, University of Pennsylvania\" />\n<meta name=\"citation_author_institution\" content=\"Childhood Cancer Data Lab, Alex\u2019s Lemonade Stand Foundation\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-8713-9213\" />\n<link rel=\"canonical\" href=\"https://jaybee84.github.io/ml-in-rd/\" />\n<meta property=\"og:url\" content=\"https://jaybee84.github.io/ml-in-rd/\" />\n<meta property=\"twitter:url\" content=\"https://jaybee84.github.io/ml-in-rd/\" />\n<meta name=\"citation_fulltext_html_url\" content=\"https://jaybee84.github.io/ml-in-rd/\" />\n<meta name=\"citation_pdf_url\" content=\"https://jaybee84.github.io/ml-in-rd/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"application/pdf\" href=\"https://jaybee84.github.io/ml-in-rd/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"text/html\" href=\"https://jaybee84.github.io/ml-in-rd/v/bb3090b12a7e63c0dda2f047a812d9e907817e83/\" />\n<meta name=\"manubot_html_url_versioned\" content=\"https://jaybee84.github.io/ml-in-rd/v/bb3090b12a7e63c0dda2f047a812d9e907817e83/\" />\n<meta name=\"manubot_pdf_url_versioned\" content=\"https://jaybee84.github.io/ml-in-rd/v/bb3090b12a7e63c0dda2f047a812d9e907817e83/manuscript.pdf\" />\n<meta property=\"og:type\" content=\"article\" />\n<meta property=\"twitter:card\" content=\"summary_large_image\" />\n<link rel=\"icon\" type=\"image/png\" sizes=\"192x192\" href=\"https://manubot.org/favicon-192x192.png\" />\n<link rel=\"mask-icon\" href=\"https://manubot.org/safari-pinned-tab.svg\" color=\"#ad1457\" />\n<meta name=\"theme-color\" content=\"#ad1457\" />\n<!-- end Manubot generated metadata -->"
keywords:
- rare disease
- machine learning
- transfer learning
lang: en-US
manubot-clear-requests-cache: false
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
title: Machine learning in rare disease
...






<small><em>
This manuscript
([permalink](https://jaybee84.github.io/ml-in-rd/v/bb3090b12a7e63c0dda2f047a812d9e907817e83/))
was automatically generated
from [jaybee84/ml-in-rd@bb3090b](https://github.com/jaybee84/ml-in-rd/tree/bb3090b12a7e63c0dda2f047a812d9e907817e83)
on March 1, 2023.
</em></small>

## Authors



+ **Jineta Banerjee**
  <sup>☯</sup><br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-1775-3645](https://orcid.org/0000-0002-1775-3645)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [jaybee84](https://github.com/jaybee84)<br>
  <small>
     Sage Bionetworks
     · Funded by Neurofibromatosis Therapeutic Acceleration Program; Children's Tumor Foundation
  </small>

+ **Jaclyn N Taroni**
  <sup>☯</sup><br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-4734-4508](https://orcid.org/0000-0003-4734-4508)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [jaclyn-taroni](https://github.com/jaclyn-taroni)<br>
  <small>
     Childhood Cancer Data Lab, Alex’s Lemonade Stand Foundation
  </small>

+ **Robert J Allaway**
  <sup></sup><br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-3573-3565](https://orcid.org/0000-0003-3573-3565)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [allaway](https://github.com/allaway)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [allawayr](https://twitter.com/allawayr)<br>
  <small>
     Sage Bionetworks
     · Funded by Neurofibromatosis Therapeutic Acceleration Program; Children's Tumor Foundation
  </small>

+ **Deepashree Venkatesh Prasad**
  <sup></sup><br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5756-4083](https://orcid.org/0000-0001-5756-4083)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [dvenprasad](https://github.com/dvenprasad)<br>
  <small>
     Childhood Cancer Data Lab, Alex’s Lemonade Stand Foundation
  </small>

+ **Justin Guinney**
  <sup></sup><br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-1477-1888](https://orcid.org/0000-0003-1477-1888)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [jguinney](https://github.com/jguinney)<br>
  <small>
     Sage Bionetworks
     · Funded by Neurofibromatosis Therapeutic Acceleration Program; Children's Tumor Foundation
  </small>

+ **Casey Greene**
  <sup>✉</sup><br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-8713-9213](https://orcid.org/0000-0001-8713-9213)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [cgreene](https://github.com/cgreene)<br>
  <small>
     Department of Systems Pharmacology and Translational Therapeutics, Perelman School of Medicine, University of Pennsylvania; Childhood Cancer Data Lab, Alex’s Lemonade Stand Foundation
  </small>

<sup>☯</sup>: These authors contributed equally to this work.

<sup>✉</sup>: Corresponding author; Please address your emails to <casey.s.greene@cuanschutz.edu>.


## Abstract {.page_break_before}

The advent of high-throughput profiling methods (such as genomics or imaging) has accelerated basic research and made deep molecular characterization of patient samples routine. 
These approaches provide a rich portrait of genes, molecular pathways, and cell types involved in disease phenotypes. 
Machine learning (ML) can be a useful tool to extract disease-relevant patterns from high dimensional datasets. 
However, depending on the complexity of the biological question, machine learning often requires a large number of samples to identify recurrent and biologically meaningful patterns. 
Rare diseases are inherently limited in clinical cases and thus have few samples to study.
In this perspective, we outline the challenges and emerging solutions for using ML in the context of small sample sets, specifically that of rare diseases. 
Advances in ML methods for rare disease are likely to be informative for applications beyond rare diseases in which sample sizes are small but datasets are high-dimensional (e.g., using genomics data for predictive modeling in precision medicine). 
We propose that the methods community prioritizes the development of ML techniques for rare disease research.


## Introduction {.page_break_before}

Rare disease researchers increasingly depend on machine learning (ML) to analyze high-dimensional datasets. 
A systematic review of ML applications in rare diseases (as defined in the European Union, i.e. fewer than 5 patients per 10,000 people) uncovered 211 human studies that used ML to study 74 different rare diseases over the last 10 years.[@doi:10.1186/s13023-020-01424-6] 
Indeed, ML can be a powerful tool in biomedical research but it does not come without pitfalls, some of which are magnified in a rare disease context.[@doi:10.3389/fmed.2021.747612]
In this perspective, we discuss considerations for using two types of ML – supervised and unsupervised learning – in the study of rare diseases, with a specific focus on high-dimensional molecular data.

ML algorithms are computational methods that identify patterns in data, often in the form of lower-dimensional representations that can be used to perform useful computational tasks.  
_Supervised learning_ algorithms must be trained with data in which samples are “labeled” with a trait of interest, such as a biological or clinical phenotype.
Supervised methods can learn correlations of features (e.g., expression measurements of a large number of genes) that may be associated with these labels to predict or infer these labels in unlabeled data, such as predicting which patients will or will not respond to treatment. 
Therefore, if a study aims to classify patients with a rare disease into disease subtypes based on high-throughput molecular profiling, a supervised ML algorithm is appropriate to carry out this task. 
Conversely, unsupervised learning algorithms learn patterns or features from unlabeled data. 
In the absence of known disease subtypes, unsupervised ML approaches can be applied to gene expression data to identify groups of samples with similar patterns of molecular states or pathway activity.  [@doi:10.1158/0008-5472.CAN-08-2100].
Unsupervised approaches can also extract combinations of features (e.g., genes) that may describe a certain cell type or pathway.
See Box 1 for more examples of how ML can be used in rare disease research.

While ML can be a useful tool, there are challenges in applying ML to rare disease datasets. 
ML methods are generally most effective when using large datasets; analyzing high dimensional biomedical data such as gene expression with many thousands of features from rare disease datasets that typically contain relatively few samples is challenging[@https://www.fda.gov/media/99546/download; @doi:10.1186/s13023-020-01424-6].
Small sample datasets tend to lack statistical power and magnify the susceptibility of ML to misinterpretation and unstable performance. 
With insufficient data, an unsupervised model will fail to identify patterns that are useful for biological discovery. 
In the case of supervised models, they can be adversely impact if sample labels are uncertain or contain “label-noise”. [@doi:10.1093/jamia/ocw028] 
Datasets with high label-noise decrease prediction accuracy and necessitate larger sample sizes during the process by which models learn patterns that distinguish samples in different classes [@doi:10.1109/tnnls.2013.2292894] (_model training_; Box 2).
Rare disease datasets often come with significant label-noise.
For example, if classifications of rare disease subtypes evolve over time, researchers constructing datasets for ML research may find that cohorts collected at different time periods do not have comparable labels.
Additionally, a supervised ML model is of limited utility if it can only accurately predict sample labels in the data it was trained on, also known as overfitting. 
Instead, most researchers aspire to develop models that generalize (maintain performance) when applied to new data that has not yet been “seen” by the model.

While we expect ML in rare disease research to continue to increase in popularity, the field requires methods that can learn patterns from small datasets and can generalize to newly acquired data [@doi:10.1016/j.ebiom.2019.08.027]. 
In this perspective, we highlight approaches that address or better tolerate the limitations of rare disease data and discuss the future of ML applications in rare disease.


## Constructing machine learning-ready rare disease datasets

High-throughput ‘omic’ assays can generate thousands to billions of measurements from whole transcriptome and whole genome sequencing respectively, resulting in high-dimensional datasets.
A typical rare disease dataset consists of a small number of samples[@doi:10.1186/s13023-020-01424-6] leading to the “curse of dimensionality” in which the feature space is much larger than the sample space, increasing the difficulty in building highly generalizable models [@doi:10.1038/nrc2294].
A larger feature space can contribute to higher data missingness (_sparsity_), more dissimilarity between samples (_variance_), and increased redundancy among individual features or combinations (_multicollinearity_) [@doi:10.1038/s41592-018-0019-x], all of which contribute to challenges in ML implementation.

An important factor in ML is model performance: the accuracy of a supervised model in identifying patterns relevant for a biological question, or the reliability of an unsupervised model in identifying hypothetical biological patterns supported by post-hoc validation.
When small sample sizes compromise an ML model’s performance, two approaches can be taken to manage sparsity, variance, and multicollinearity: 1) increase the number of samples, 2) improve the quality of samples.
In the first approach, appropriate training, evaluation, and held-out validation sets could be constructed by combining multiple rare disease cohorts (Figure [@fig:1]a, Box 2).
When combining datasets, special attention should be directed towards data harmonization since data collection methods can differ between cohorts.
Without careful selection of aggregation methods, one may introduce technical – in contrast to biological – variability into the combined dataset and negatively impact the ML model’s ability to learn or detect meaningful signals.
Steps like reprocessing data using a single pipeline, using batch correction methods [@doi:10.1093/biostatistics/kxj037; @doi:10.1093/nar/gku864], and normalizing raw values appropriately without affecting the underlying variance in the data [@doi:10.1186/gb-2010-11-3-r25, @doi:10.1371/journal.pcbi.1003531, @doi:10.1186/s13059-014-0550-8] may be necessary to mitigate unwanted variability. (Figure [@fig:1]a)
Data harmonization may also entail standardization of sample labels using biomedical ontologies to normalize how samples are described across multiple datasets.

How does one know if a composite dataset has undergone proper harmonization and annotation?
Ideally, the dominant patterns of the composite dataset reflect variables of interest, such as phenotype labels rather than technical labels.
In the latter case, this suggests that the datasets used to generate the composite dataset need to be corrected to overcome differences in how the data were generated or collected.
In the next section, we discuss approaches that help identify and visualize structure in datasets to determine whether composite rare disease datasets are appropriate for ML use.


## Learning representations from rare disease data

Dimensionality reduction methods help explore and visualize underlying structure in the data (e.g., [@doi:10.1038/s41467-019-13056-x]), to define sample subgroups (e.g., [@doi:10.1038/s41467-020-15351-4]), or for feature selection and extraction during application of specific machine learning models [@doi:10.1007/978-3-030-03243-2_299-1] (Figure [@fig:2]c).
Unsupervised methods, in finding low-dimensional patterns in data, can ‘compress’ information from a large number of features into a smaller number of features [@doi:10.1007/978-3-540-33037-0; @doi:10.1098/rsta.2015.0202, @https://www.jmlr.org/papers/v9/vandermaaten08a.html; @https://arxiv.org/abs/1802.03426] (Figure [@fig:2]).
A method commonly used for dimensionality reduction is principal components analysis (PCA).
PCA identifies higher order features, termed _principal components_ (PCs), that are combinations of original features.
The PCs are calculated in a way that maximizes the amount of information (variance) they contain and ensures that each PC is uncorrelated with the other PCs. [@doi:10.1098/rsta.2015.0202]
In practice, researchers often use the first few PCs to reduce the dimensionality without removing what may be important biological variability in the data.
Nguyen and Holmes highlight the use of "elbow method" to select the number of appropriate dimensions. [@doi:10.1371/journal.pcbi.1006907]
Multidimensional scaling (MDS), t-distributed stochastic neighbor embedding (t-SNE), and uniform manifold approximation and projection (UMAP) are other popular dimension reduction methods, often used for low-dimensional visualization and interpretation of data [@https://arxiv.org/abs/1802.03426; @doi:10.23915/distill.00002]
Testing multiple dimensionality reduction methods may be necessary to obtain a more comprehensive portrait of the data. [@doi:10.1186/s13059-020-02021-3]
Other unsupervised learning approaches such as k-means or hierarchical clustering are used to characterize structure in genomic and imaging data. [@doi:10.1186/1471-2105-9-497; @doi:10.1109/jbhi.2013.2276766]
Dimensionality reduction methods are a subset of a type of ML called representation learning.
Representation learning methods have been used to extract features from transcriptomics datasets made of combinations of gene expression values [@doi:10.1038/s41467-020-14666-6; @doi:10.1093/bioinformatics/btq503; @doi:10.1186/s13059-020-02021-3], predict rare pathologies from images [@doi:10.1016/j.media.2020.101660] (Box 1a) or detect cell populations associated with rare diseases [@doi:10.1038/ncomms14825].

When applied to complex systems, representation learning generally requires many samples and therefore may appear to aggravate the curse of dimensionality.
However, it can be a powerful tool to learn low-dimensional patterns from large datasets and then find those patterns in smaller, related datasets.
In later sections, we discuss this method of leveraging large datasets to reduce dimensionality in smaller datasets, also known as feature-representation-transfer learning.
Once the dimensions of the training dataset have been reduced, model training can proceed using the experimental design as outlined in Box 2.


## Reducing misinterpretation of model output with statistical techniques

The successful application of ML can be improved by meeting certain conditions.
First, the dataset contains sufficient representation from each class such that relevant variability from that class is captured.
Second, the dataset is complete; all samples have measurements for all variables in the dataset (i.e., the dataset is not “sparse”, it is not missing data for some of the samples).
Third, there is no ambiguity about the labels for the samples in the dataset (i.e., no “label-noise”).

Rare disease datasets violate many of these assumptions.
Small number of samples for specific classes fail to fully capture the sample variability in those classes, e.g. only a few patients with a particular rare disease in a health records dataset, which can require special consideration for evaluation (Box 2).
The data are also often sparse, and there may be abundant label-noise due to incomplete understanding of the disease.
All of these contribute to low signal to noise ratio in rare disease datasets.
Applying ML to such data without addressing these shortcomings may lead to models that have poor generalizability or are hard to interpret.

Class imbalance in datasets can be addressed using decision tree-based ensemble learning methods (e.g., random forests). [@doi:10.1007/s11634-019-00354-x] (Figure[@fig:3]a)
Random forests use sampling with replacement based techniques to form a consensus about the important predictive features identified by the decision trees (e.g., Box 1c). [@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134]
Additional approaches like combining random forests with sampling without replacement can generate confidence intervals for the model predictions (for applications like Box 1d) by mimicking real world cases where most rare disease datasets are incomplete [@doi:10.3390/genes11020226].
Resampling approaches are most helpful in constructing confidence intervals for algorithms that generate the same outcome every time they are run (i.e., deterministic models).
For decision trees that choose features at random for selecting a path to the outcome (i.e., are non-deterministic), resampling approaches can be helpful in estimating the reproducibility of the model.

In situations where decision tree-based ensemble methods fail when applied to rare disease datasets, cascade learning is a viable alternative. [@pmc:PMC6371307]
In cascade learning, multiple methods leveraging distinct statistical techniques are used to identify stable patterns in the dataset [@https://doi.org/10.1109/CVPR.2001.990537; @doi:10.1007/978-3-540-75175-5_16; @doi:10.1109/icpr.2004.1334680].
For example, a cascade learning approach for identifying rare disease patients from electronic health record data (Box 1a) incorporated independent steps for feature extraction (word2vec [@arxiv:1301.3781]), preliminary prediction with ensembled decision trees, and then prediction refinement using data similarity metrics. [@pmc:PMC6371307]
Combining these three methods resulted in better overall prediction when implemented on a silver standard dataset, as compared to a model that used ensemble-based prediction alone.
In addition to cascade learning, approaches that better represent rare classes using class re-balancing techniques like inverse sampling probability weighting [@doi:10.1186/s12911-021-01688-3], inverse class frequency weighting [@doi:10.1197/jamia.M3095], oversampling of rare classes [@https://doi.org/10.1613/jair.953], or uniformly random undersampling of majority class [@doi:10.48550/arXiv.1608.06048] may also help minimize issues associated with class imbalance.

The presence of label-noise and sparsity in the data can lead to poor generalizability or overfitting, meaning that the models show high prediction accuracy on the training data but low prediction accuracy on new evaluation data.
Overfit models tend to rely on patterns that are unique to the training data, such as the clinical coding practices at a hospital, and not generalize to new data such as data collected at different hospitals. [@isbn:0262035618; @pmc:PMC8238368]
Regularization approaches can help mitigate these scenarios by adding constraints to a model to avoid making large prediction errors.
This protects ML models from poor generalizability by reducing model complexity and minimizing model feature space [@doi:10.1371/journal.pgen.1004754, @doi:10.1002/sim.6782]. (Figure[@fig:3]a)
Examples of ML methods with regularization include ridge regression, LASSO regression, and elastic net regression [@doi:10.1111/j.1467-9868.2005.00503.x], among others.
LASSO regularization helped select a few informative genes as features to include in models classifying amyotrophic lateral sclerosis (ALS) patients and healthy patients with high accuracy based on brain tissue gene expression, thus making the models more interpretable. [@doi:10.1186/s10020-023-00603-y]
In the context of rare immune cell signature discovery, where a few genes or features are expected to distinguish between immune cell types, elastic-net regression was able to exclude groups of uninformative genes by reducing their contribution to zero. [@doi:10.1186/s12859-019-2994-z, @https://doi.org/10.1111/j.1467-9868.2005.00503.x]
In a study using a variational autoencoder (VAE) (see Box 3) for dimensionality reduction in gene expression data from acute myeloid leukemia (AML) samples, the KL loss between the input data and its low dimensional representation provided the regularizing penalty for the model. [@doi:10.1101/278739; @doi:10.48550/arXiv.1312.6114]
A study using a convolutional neural network (CNN) to identify tubers in MRI images from tuberous sclerosis patients (an application that can facilitate Box 1a), minimized overfitting using the dropout regularization method which removed randomly chosen network nodes in each iteration of the CNN model generating simpler models in each iteration.[@doi:10.1371/journal.pone.0232376]
Thus, depending on the learning method used, regularization approaches should be considered when working with rare disease datasets.

### Building upon prior knowledge and indirectly related data {.page_break_before}

One strategy to overcome the paucity of data in rare disease is to combine a variety of data types and explore rare disease data alongside other existing knowledge.
By using several data modalities, such as curated pathways, genetic data, or drug-target relationships, it may be possible to gain a better understanding of rare diseases.
Knowledge graphs (KGs) which integrate related-but-different data types, provide a rich multimodal data source e.g., Monarch Graph Database [@doi:10.1093/nar/gkw1128], hetionet [@doi:10.7554/elife.26726], PheKnowLator [@doi:10.1101/2020.04.30.071407], and the Global Network of Biomedical Relationships [@doi:10.1093/bioinformatics/bty114], Orphanet [@http://www.orpha.net].
These graphs connect genetic, functional, chemical, clinical, and ontological data so that relationships of data with disease phenotypes can be explored through manual review [@doi:10.1093/database/baaa015] or computational methods [@doi:10.1016/j.jbi.2021.103838, @doi:10.1142/9789811215636_0041; @doi:10.1186/s12911-019-0938-1]. (Figure[@fig:3]a)
KGs may include links (i.e. edges) or nodes that are specific to a rare disease of interest (e.g., an FDA approved treatment would be a specific disease-compound edge in the KG) and more generalized information (e.g., gene-gene interactions noted in the literature for a different disease). (Figure [@fig:4]a)

Rare disease researchers can repurpose general biological or chemical KGs that are not disease-specific to answer rare disease-based research questions [@doi:10.1142/9789811215636_0041] (e.g. Box 1b).
One tactic to sift through the large amounts of data encoded in KGs is to calculate the distances between nodes of interest (e.g., diseases and drugs for Box 1b [@doi:10.1142/9789811215636_0041]); often done by determining the "embeddings" - lower dimension vector representations of the position and connections of a particular point in the graph for nodes in the KG, and calculating the similarity between these embeddings.
Effective methods to calculate node embeddings that can generate actionable insights for rare diseases is an active area of research [@doi:10.1142/9789811215636_0041].

Another application of KGs is to augment a dataset [@doi:10.1186/s12911-019-0752-9, doi:10.1186/s12911-019-0938-1].
Li et. al.[@doi:10.1186/s12911-019-0938-1] used a KG to identify linked terms in a medical corpus from a large number of patients, some with rare disease diagnoses.
They augmented their text dataset by mapping related clinical terms together - e.g., mapping "cancer" and "malignancy" in different patients to the same clinical concept.
With this enhanced dataset, they trained and tested a variety of text classification algorithms to identify rare disease patients within their corpus. (Figure [@fig:4]b, Box 1a)

Rare disease researchers have also integrated multiple KGs and applied neural network-based algorithms optimized for graph data, such as a graph convolutional neural networks.
Rao and colleagues [@doi:10.1186/s12920-018-0372-8] describe the construction of a KG using phenotype information from the Human Phenotype Ontology, and rare disease information from Orphanet and curated gene interaction/pathway data from Lit-BM-13 and WikiPathways [@pmc:PMC7778952; @doi:10.1016/j.cell.2014.10.050; @doi:10.1093/nar/gkaa1024].
They trained a spectral graph convolution neural network on this KG to identify and rank potentially causal genes for the rare diseases from Orphanet, and used this model to accurately predict causal genes for a ground truth dataset of rare diseases with known causal genes.
While several groups have used KGs to study rare diseases, we expect that better multi-modal datasets and ML methods to analyze KGs will make them a more popular and important tool in rare disease.

Another approach that builds on prior knowledge and large volumes of related data is transfer learning, a modeling technique that “borrows strength” across datasets with both similar and distinct properties like an imaging anomaly present in both rare and common diseases, to advance our understanding of rare diseases.
Transfer learning, where a model trained for one task or domain (source domain) is applied to another related task or domain (target domain), can be supervised or unsupervised.
Among various types of transfer learning, feature-representation-transfer approaches learn representations from the source domain and apply them to a target domain [@doi:10.1109/tkde.2009.191](Figure [@fig:5]a-c).
That is, representation learning, as discussed earlier, does not need to be applied only to describe the dataset on which the algorithm was trained – it can also be used to elucidate signals in sufficiently similar data (Figure [@fig:5]c) and may offer an improvement in descriptive capability over models trained only on small rare disease datasets (Fig [@fig:5]c).
For instance, low-dimensional representations can be learned from tumor transcriptomic data and transferred to describe patterns in genetic alterations in cell lines [@doi:10.1186/s13059-020-02021-3](Figure [@fig:5]c).
In the next section, we summarize specific instances of applying transfer learning, along with other techniques, to the study of rare diseases.


### Combining approaches is required for the successful application of machine learning to rare diseases

We have described multiple approaches for maximizing the success of ML applications in rare disease, but it is rarely sufficient to use any of these techniques in isolation.
Below, we highlight two examples in the rare disease domain that use concepts of feature-representation-transfer, use of prior data, and regularization.

Our first example includes a large dataset of acute myeloid leukemia (AML) patient samples with no drug response data and a small _in vitro_ experiment with drug response data [@doi:10.1038/s41467-017-02465-5].
Training an ML model on the small _in vitro_ dataset alone faced the _curse of dimensionality_ and the dataset size prohibited representation learning.
Dincer et al. trained a variational autoencoder (VAE, Box 3) on a reasonably large dataset of AML patient samples from 96 independent studies to learn meaningful representations in an approach termed DeepProfile [@doi:10.1101/278739] (Figure[@fig:6]a).
The representations or _encodings_ learned by the VAE were then _transferred_ to the small _in vitro_ dataset reducing its number of features from thousands to eight, and improving the performance of the final LASSO linear regression model (Box 1b).
In addition to improving performance, the _encodings_ learned by the VAE captured more biological pathways than PCA, possibly due to the constraints on the encodings imposed during training (Box 3).
Similar results were observed for prediction of histopathology in another rare cancer dataset [@doi:10.1101/278739].

While DeepProfile was centered on training on an individual disease and tissue combination, some rare diseases affect multiple tissues that a researcher may want to study (e.g., Box 1d).
Studying multiple tissues poses significant challenges and a cross-tissue analysis may require comparing representations from multiple models.
Models trained on a low number of samples may learn representations that "lump together" multiple biological signals, reducing the interpretability of the results.
To address these challenges, Taroni et al. trained a Pathway-Level Information ExtractoR (PLIER) (a matrix factorization approach that takes prior knowledge in the form of gene sets or pathways) [@doi:10.1038/s41592-019-0456-1] on a large generic collection of human transcriptomic data [@doi:10.1016/j.cels.2019.04.003].
PLIER used constraints (regularization) that learned _latent variables_ aligned with a small number of input gene sets, making it suitable for rare disease data.
The authors _transferred_ the representations or _latent variables_ learned by the model to describe transcriptomic data from the unseen rare diseases antineutrophil cytoplasmic antibody (ANCA)-associated vasculitis (AAV) and medulloblastoma in an approach termed MultiPLIER [@doi:10.1016/j.cels.2019.04.003]. (Figure[@fig:6]b)
MultiPLIER used one model to describe multiple datasets instead of reconciling output from multiple models, making it possible to identify commonalities among disease manifestations or affected tissues.

DeepProfile [@doi:10.1101/278739] and MultiPLIER [@doi:10.1016/j.cels.2019.04.003] exemplify modeling approaches incorporating prior knowledge – thereby constraining the model space according to plausible or expected biology – or sharing information across datasets.
These two methods capitalize on similar biological processes observed across different biological contexts and the fact that the methods underlying the approaches can effectively learn about those processes.


## Outlook

This perspective highlights challenges in applying ML to rare disease data and approaches that address these challenges.
Small sample size, while significant, is not the only roadblock.
The high dimensionality of modern data requires creative approaches, such as learning new representations of the data, to manage the curse of dimensionality.
Leveraging prior knowledge and transfer learning methods to appropriately interpret data is also required.
Furthermore, we posit that researchers applying ML methods on rare disease data should use techniques that increase confidence (i.e., bootstrapping) and penalize complexity of the resultant models (i.e., regularization) to enhance the generalizability of their work.
It should be noted that the line between classical statistical methods and ML is fuzzy.
Multiple statistical techniques that were considered to be out of scope of this article (e.g. hierarchical models, Bayesian frameworks, association tests) [@doi:10.1016/j.ajhg.2016.01.008; @doi:10.1016/j.ajhg.2011.05.029; @doi:10.1371/journal.pgen.1004729; @doi:10.1016/j.ajhg.2017.05.015], may have substantial potential to enhance the accuracy and generalizability of models, and should be considered in the rare disease study design process.

The approaches highlighted in this perspective come with challenges that may undermine investigators' confidence in using these techniques for rare disease research.
We believe that the challenges in applying ML to rare disease are opportunities to improve data generation and method development going forward.
The following two areas are particularly important for the field to explore.

### Intentional data generation and sharing mechanisms are key for powering the future of rare disease data analysis

While many techniques exist to collate rare data from different sources, low-quality data may hurt the end goal even if it increases the size of the dataset.
In our experience, collaboration with domain experts has proved to be critical in gaining insight into potential sources of variation in the datasets.
An anecdotal example: conversations with a clinician revealed that samples in a particular tumor dataset were collected using vastly different surgical techniques (laser ablation and excision vs standard excision).
This information, not readily available to non-experts, was obvious to the clinician.
Such instances suggest that collaboration with domain experts and sharing of well-annotated data is needed to generate robust datasets.

In addition to sample scarcity, comprehensive phenotypic-genotypic databases are also lacking.
While rare disease studies that collect genomic and phenotypic data are becoming more common [@doi:10.1038/nrg3555; @doi:10.1038/nrg.2017.116; @doi:10.1056/NEJMra1711801], developing comprehensive genomics-based genotype-phenotype databases that prioritize clinical and genomics data standards is key to fueling interpretation of ML methods.
This method can be bolstered by funding or fostering collaboration between biobanking projects and patient registry initiatives.
Mindful sharing of data with proper metadata and attribution enabling prompt data reuse is important in building valuable datasets for rare disease research [@https://www.nature.com/articles/s41576-020-0257-5].
Finally, federated learning methods, such as those used in mobile health [@doi:10.1038/s41746-020-00323-1] and electronic healthcare records studies [@doi:10.1001/jamanetworkopen.2021.24946], may allow researchers to develop ML models on data from larger numbers of people with rare diseases whilst protecting patient privacy.

### Methods that reliably support mechanistic interrogation of specific rare diseases are an unmet need

Most ML methods for rare diseases are used for classification tasks.
Not many methods investigate biological mechanisms; this is likely due to a lack of methods that can handle the limitations of rare disease data described throughout this perspective.
Developing methods to address this will be critical for applying ML to rare disease data.

For example, development of methods with a focus on explainability of the model can identify features that may be related to the underlying disease mechanism.[@doi:10.1038/s41551-018-0304-0]
Alternatively, representation learning or regularization methods may help identify multiple correlated features which can be interrogated to identify biologically meaningful pathways.
Additionally, robust _error analysis_ for newly developed models to help users understand how a feature influences the performance of a model can provide insight into its potential contribution to the underlying disease mechanism. [@doi:10.1093/bioinformatics/bth060]
Interrogating disease mechanisms by adopting modifications of these approaches is necessary as ML applications become mainstream in research and clinical settings.

Finally, methods that can reliably integrate disparate datasets will likely always remain a need in rare disease research.
Methods that rely on finding structural correspondence between datasets ("anchors") may be able to transform the status-quo of using ML in rare disease [@https://www.aclweb.org/anthology/W06-1615; @https://dl.acm.org/doi/10.5555/2283516.2283652; @doi:10.1016/j.cell.2019.05.031].
We speculate that this is an important and burgeoning area of research, and we are optimistic about the future of applying ML approaches to rare diseases.

## Ethics Declaration

Justin Guinney is currently employed at Tempus Labs, a precision medicine company.
Jaclyn N. Taroni and Deepashree Venkatesh Prasad are employed with Alex's Lemonade Stand Foundation, a research funder.




## Author Contributions

Authorship was determined using ICMJE recommendations.
Conceptualization - J.B., J.N.T, R.J.A, C.G., J.G
Data curation - Not applicable 
Formal Analysis - Not applicable 
Funding acquisition - R.J.A, 
Investigation - J.B., J.N.T, R.J.A; 
Methodology - J.B., J.N.T, R.J.A.; 
Project administration - J.B.
Resources - J.B., J.N.T, R.J.A; 
Software - Not applicable 
Supervision - J.B., C.G 
Validation - Not applicable 
Visualization - D.V.P 
Writing – original draft - J.B., J.N.T., R.J.A.;
Writing – review & editing- J.B., J.N.T., R.J.A., C.G., J.G.


## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs"></div>


## Figure Legends

![Combining datasets to increase data for training machine learning models. a) Appropriate methods are required to combine smaller datasets into a larger composite dataset: The left panel shows multiple small, rare disease datasets that need to be combined to form a dataset of higher sample size. The color of the samples suggests classes or groups present in the datasets. The shape represents the dataset of origin. The middle panel shows methods that may be used to combine the datasets while accounting for dataset-specific technical differences. The right panel shows Principal Component Analysis of the combined datasets to verify proper integration of samples in the larger dataset. b) Composite datasets can be used to make training, evaluation, and validation datasets for machine learning: Left panel shows the division of the composite dataset into training dataset and a held-out validation dataset (top). The held-out validation set is a separate study that has not been seen by the model. The training set is further divided into training and evaluation datasets for k-fold cross-validation (in this example k=4), where each fold contains all samples from an individual study. This approach is termed study-wise cross validation and supports the goal of training models that generalize to unseen cohorts. c) Barplot showing the class distribution of the training, evaluation, and held-out validation datasets from panel (b).](images/figures/lower-res-figures/figure-1-combining-datasets.png){#fig:1}

![Representation learning can extract useful features from high dimensional data. a) The data (e.g., transcriptomic data) are highly dimensional, having thousands of features (displayed as Fa-Fz). Samples come from two separate classes (purple and green row annotation).  b) In the original feature space, Fa and Fb do not separate the two classes (purple and green) well. c) A representation learning approach learns new features (e.g., New Feature 1, a combination of Fa, Fb .... Fz, and New Feature 2, a different combination of Fa, Fb .... Fz). New Feature 2 distinguishes class, whereas New Feature 1 may capture some other variable such as batch (not represented). New features from the model can be used to interrogate the biology of the input samples, develop classification models, or use other analytical techniques that would have been more difficult with the original dataset dimensions.](images/figures/pdfs/figure2-representation-learning-resized.png){#fig:2}

![Strategies to reduce misinterpretation of machine learning model output in rare disease. a) Bootstrapping: Left panel shows a small rare disease dataset, which can be resampled with replacement using bootstrap to form a large resampled dataset (middle panel). Running the same ML model on multiple resampled datasets generates a distribution of values for the importance scores for each feature utilized by the ML model (right panel), b) Cascade Learning: A schematic showing the different steps in a cascade learning approach for identifying rare disease patients from electronic health record data. The bar plot in the middle panel schematically represents patient classification accuracy after ensemble learning. The accuracy is high for non-rare diseases, but low for rare diseases. The bar plot on the right panel depicts classification accuracy after implementation of cascade learning. The accuracy is high for both non-rare and rare diseases. c) Regularization: A schematic showing the concept of regularization to selectively learn relevant features. The samples (green and blue circles) in the rare disease dataset on left panel can be represented as a combination of features. Each horizontal bar in the middle panel (training set) represents a feature-by-sample heatmap for one sample each. In the held-out validation dataset, for a sample of unknown class (open circle), some features recapitulate the pattern present in the training set, while others do not. The right panel depicts accuracy of predicting the class of the open circles with or without using regularization during implementation of the ML models on rare disease data. Without regularization the classification accuracy is low due to presence of only a subset of learned features (denoted by dashed rectangle in middle panel), but with regularization this subset of features is sufficient to gain high classification accuracy.](images/figures/pdfs/figure-3-resized.png){#fig:3}

![Application of knowledge graphs can improve machine learning in rare disease. a) Knowledge graphs integrate different data types (e.g., genetic, functional, clinical, chemical, and ontological data) and may allow models to learn from connections that are rare disease-specific or happen in many biomedical contexts. There are a variety of possible applications of this approach, including identifying new disease-drug relationships [@doi:10.1142/9789811215636_0041], augmenting data to improve accuracy of models trained on the data [@doi:10.1186/s12911-019-0752-9, doi:10.1186/s12911-019-0938-1], or mining prior knowledge to discover important gene sets and pathways in rare diseases [@doi:10.1186/s12920-018-0372-8].  b) Knowledge graphs can also be used to augment data. Li et. al. [@doi:10.1186/s12911-019-0938-1] applied a classifier to an EHR corpus to identify rare disease patients. They trained a classifier on the EHR data alone (e.g., thrombocytopenia, anemia) and trained another classifier on data augmented with medically-related concepts from a knowledge graph (e.g., neutropenia, stroke). The classifier trained on knowledge-graph augmented data has lower error and higher accuracy (right panel). ](images/figures/pdfs/figure-4-KG.png){#fig:4}

![Feature-representation-transfer approaches learn representations from a source domain and apply them to a target domain. a) Combination of features representing samples of a large dataset (transcriptomic data from tumors) are learned by an ML model through representation learning.  b) When applied to a small cell line dataset, the representations extracted by an ML model tend to be incomplete and correlate poorly with clinical or drug sensitivity features. c) When a representation learning model trained on the large dataset (a) is applied to the small cell line dataset to extract consistent combinations of features based on the combinations found in the larger training dataset, the extracted representations correlate strongly with the clinical or drug sensitivity features ](images/figures/pdfs/figure-5-transfer-learning.png){#fig:5}

![Combining multiple strategies strengthens the performance of ML models in rare disease. a) The authors of DeepProfile trained a variational autoencoder (VAE) to learn a representation from acute myeloid leukemia data without phenotype labels, transferred those representations to a small dataset with phenotype labels, and found that it improved prediction performance in a drug sensitivity prediction task [@doi:10.1101/278739]. b) The authors of MultiPLIER trained a Pathway-Level Information ExtractoR (PLIER) model on a large, heterogeneous collection of expression data (recount2 [@doi:10.1038/nbt.3838]) and transferred the representations (termed latent variables) to multiple datasets from rare diseases that were not in the training set [@doi:10.1038/s41592-019-0456-1]. Expression of PLIER latent variables can be used to check for concordance between datasets, among other applications.](images/figures/pdfs/figure-6.png){#fig:6}


## Box 1: Common uses for machine learning in rare disease:

### (a) Identifying patients with rare diseases

ML can be used to identify features in high dimensional data that correlate strongly with a patient or sample phenotype and subsequently predict the presence or absence of a rare disease. 
For example, supervised ML models can be trained on electronic health records, genetic data, or medical images to identify potential new patients with a rare disease.

### (b) Drug discovery or repurposing 

ML can help identify potential drug candidates for rare diseases. 
For example, unsupervised and supervised algorithms trained on genetic and molecular data from high throughput screens can identify novel therapeutic targets for a rare disease.
Additionally, algorithms utilizing knowledge graphs, genomic data, and databases of existing approved drugs can identify potential therapeutic candidates for rare diseases.

### (c) Clinical trial design improvement

Optimized study design and identification of appropriate trial participants can greatly reduce costs while increasing the likelihood of successful outcomes for clinical trials.
ML approaches can benefit clinical trial study design.
For example, unsupervised ML approaches can be used to identify sub-groups of patients who are more likely to respond well to a particular treatment.
Supervised ML approaches can also be used to predict drug response in rare disease patients.

### (d) Molecular subtyping of disease

Rare diseases often show overlapping and heterogenous phenotypes.
ML approaches can be used to identify molecular subtypes of the disease for better understanding.
For example, unsupervised ML approaches can help identify new subtypes of a rare disease using molecular and genetic data.
The same approaches can help identify the important molecular features that define the subtypes.

### (e) Patient prognosis prediction

Rare diseases can suffer from lack of in-depth understanding of disease mechanism. 
Biomarkers or clinical features that correlate strongly with adverse outcomes can be beneficial in predicting prognosis of a patient.
Supervised ML algorithms can be useful in identifying factors contributing to risk of adverse outcomes or progression to advanced disease in rare disease patients.
Patient stratification can help identify patient subpopulations who can benefit by early and aggressive interventions.


## Box 2: Understanding experimental design of ML to inform requirements for data:

### Components of ML experiments

ML algorithms identify patterns that explain or fit a given dataset.
Every ML algorithm goes through _training_, where it identifies underlying patterns in a given dataset to create a “trained” algorithm (a model), and _testing_, where the model applies the identified patterns to unseen data points. 
Typically, a ML algorithm is provided with: 1. a _training dataset_ , 2. an _evaluation dataset_ , 3. a _held-out validation dataset_. 
These input data can be images, text, numbers, or other types of data which are typically encoded as a numerical representation of the input data. 
A training dataset is used by the model to learn underlying patterns from the features present in the data of interest. 
An evaluation dataset is a small and previously unused dataset which is used during the training phase to help the model iteratively update its parameters (i.e., _hyperparameter tuning_ or _model tuning_). 
In many cases, a large training set may be subdivided to form a smaller training dataset and the evaluation dataset, both of which are used to train the model. 
In the testing phase, a completely new or unseen test dataset or held-out validation set is used to test whether the patterns learned by the model hold true in new data (i.e., they are _generalizable_). 
While the evaluation dataset helps us refine a model’s fit to patterns in the training data, the held-out validation set helps us test the generalizability of the model.

If a model is generalizable, it is able to make accurate predictions on new data.
High generalizability of a model on previously unseen data suggests that the model has identified important patterns in the data that are not unique to the data used for training and tuning.
Generalizability can be affected if _data leakage_ occurs during training of the model, i.e., if a model is exposed to the same or related data points in both the training set and the held-out validation set.
Ensuring absence of any overlap or relatedness among data points or samples used in the training set and evaluation set is important to avoid data leakage during model training.
Specifically, in cases of rare genetic diseases where, for example, many samples can contain familial relationships or data from the same patient could be collected by multiple specialists at different clinical facilities, special care should be taken while crafting the training and testing sets to ensure that no data leakage occur and the trained model has high generalizability.

### Training and testing

The implementation of a ML experiment begins with splitting a single dataset of interest such that a large proportion of the data is used for training, and the remaining data is used for testing or validation as the held-out validation dataset.
The training dataset is generally subdivided into the training dataset and the evaluation dataset.
Ideally, a _held-out validation dataset_ is an entirely new study or cohort, as researchers typically aim to build models that generalize to unseen, newly generated data. 
In rare diseases where multiple datasets may be combined to make a large enough training dataset, special care should be taken to standardize the features and the patterns therein. 
Although ML algorithms generally expect that datasets have uniform features, normalizing training and testing data together may introduce similarities between samples (causing inadvertent data leakage) that hamper the goal of training models that are highly generalizable.

The iterative training phase helps the model learn important patterns in the training dataset and then use the evaluation dataset to test for errors in prediction and update its learning parameters (hyperparameter tuning). 
The method by which the trained model is applied to the evaluation dataset to measure performance and update the hyperparameters is called cross-validation. 
There are multiple approaches that can be deployed to maximally utilize the available data when generating training and evaluation datasets e.g., leave-p-out cross-validation, leave-one-out cross-validation, k-fold cross-validation, Monte-Carlo random subsampling cross-validation.[@doi:10.1007/978-1-4614-6849-3]
In the case of k-fold cross-validation, a given dataset is shuffled randomly and split into _k_ parts. 
One of the k parts is reserved as the _evaluation dataset_ and the rest are combined and used as the _training dataset_. 
In the next iteration, a different part is used as the evaluation dataset, while the rest are used for training. 
To avoid data leakage, and to promote generalization of models to new studies, researchers can use _study-wise cross-validation_, such that all samples from a study are in the same fold and no individual study is represented in both the training and evaluation datasets. 
Once the model has iterated through all k parts of the training and evaluation datasets, it is ready to be tested on the held-out validation dataset.(Figure [@fig:1]b)

The held-out validation dataset is exposed to the model only once to estimate the accuracy of the model.
High accuracy of a model during cross-validation but low accuracy on the held-out validation dataset is a sign that the model has become overfit to the training set and has low generalizability.
If there is evidence of overfitting, researchers should revisit the construction of the dataset to make sure they meet the best practices outlined above.

It is important to note that accuracy alone may not be the best measure of performance in rare disease datasets.
A model tested for identifying rare disease samples may still achieve high accuracy if it identifies every sample as a non-rare disease sample.
Measures that are more suitable to handle class imbalance, such as the Kappa statistic or area under the precision-recall curve [@doi:10.1145/1143844.1143874], are better metrics for model performance for rare disease. 

## Box 3: Definitions: {.page_break_before}

### Knowledge Graph

A knowledge graph is a network representation of human knowledge about a domain, abstracted into nodes and edges. 
Any entity of interest (for example a gene, a disease, a protein, or a cell-line) can be represented as a node in a knowledge graph.
All nodes can be linked through edges that represent known relationships between the nodes. 
Edges can be directed, indicating that the order of the nodes is important for encoding the relationship, or undirected.
For example, a gene (node) can be linked to a protein (node) using a directed edge that represents the relationship that the protein is generated through the transcription and translation of the gene. 
Knowledge graphs serve to integrate data that exist in distributed sources, encode human readable knowledge in machine readable format, and evolve in a flexible manner to integrate new knowledge as it becomes available.

### Machine Learning

Machine learning is a scientific discipline at the intersection of computer science and statistics, which combines computational and statistical methods to identify patterns in sample data.[@https://link.springer.com/book/10.1007/978-0-387-21606-5, @doi:10.1038/s41580-021-00407-0]
In this discipline, one intends to use data as input and apply or fit predictive models to recognize patterns in the data or identify informative groups among the data using objective computational methods.

### Rare Disease

According to the Orphan Drug Act[@https://www.fda.gov/media/99546/download] of United States of America, diseases or conditions that impact less than 200,000 people in the U.S are considered to be rare diseases. 
The European Union defines a disease as rare when it affects less than 1 in 2,000 people.

### Regularization
Regularization is an approach to reduce overfitting of models to training data, where a penalty or constraint is added to a model trained on a training dataset to avoid making large prediction errors on the evaluation dataset.

### Transfer Learning

Transfer learning is an approach where a model trained for one task or domain (source domain) is applied to another, typically related task or domain (target domain), for example a model pre-trained natural images from the ImageNet dataset can potentially be used to classify medical images.[@doi:10.1109/TMI.2016.2528162]
Transfer learning can be supervised (one or both of the source and target domains have labels), or unsupervised (both domains are unlabeled).

### Variational Autoencoder

Variational Autoencoders or VAEs are unsupervised neural networks that use hidden layers to learn or encode representations from available data while mapping the input data to the output data. 
VAEs are distinct from other autoencoders since the distribution of the encodings are regularized such that they are close to a normal distribution, which may contribute to learning more biologically relevant signals [@doi:10.1186/s13059-020-02021-3].

# Classification in the Presence of Label Noise: a Survey

[@doi:10.1109/tnnls.2013.2292894]

Frénay and Verleysen (2014) define and describe sources of label noise, consequences of label noise on ML model development, and techniques that can help mitigate the resulting issues. 
It also suggests few quality metrics that may be used to test newly developed models to understand their performance in presence of noisy data.

# Random Forests

[@doi:10.1023/a:1010933404324]

Breiman (2007) describes the theory underlying random forest algorithm which is a popular ML method with applications in almost every high-dimensional data field, including medicine and life sciences. 
The paper also highlights the behavior of random forests in the presence of data where each individual feature is not strong enough to distinguish between classes (weak inputs). 
This scenario is frequently encountered in biomedical data and especially rare disease data.

# Deriving disease modules from the compressed transcriptional space embedded in a deep autoencoder

[@doi:10.1038/s41467-020-14666-6]

Dwivedi, et. al. (2020) apply representation learning to meaningfully group genes into modules in a data-driven manner. 
The authors turned the deep neural network into a "glass-box" by layering knowledge-based interpretation on the gene modules identified using a variational autoencoder to provide insights into what each layer of their model represent.

# Pathway-level information extractor (PLIER) for gene expression data

[@doi:10.1038/s41592-019-0456-1]

Mao and colleagues (2019) use the principles of representation learning combined with prior-knowledge to develop a model that capitalizes on correlated gene expression to identify meaningful biological pathways and reduce noise in data introduced due to technical variation.
This method has been the foundation of mutiple recent works applying ML on rare disease gene expression data.

# Automatic detection of rare pathologies in fundus photographs using few-shot learning

[@doi:10.1016/j.media.2020.101660]

Quellec, et. al. (2020) train a convolutional neural network (CNN) to identify rare pathologies in retinal images.
They used an approach called few-shot learning, in which the CNN is trained on a large cohort of retinal images and then trained on a very small number of examples of retinal images with rare pathologies.

# A Literature-Based Knowledge Graph Embedding Method for Identifying Drug Repurposing: Opportunities in Rare Diseases

[@doi:10.1142/9789811215636_0041]

Sosa and colleages (2020) develop a knowledge graph of drug, disease, and gene relationships from pubished medical literature. 
They then use a graph embedding method to explore the knowledge graph and predict drug repurposing opportunities. 
