### Techniques that build on prior knowledge and indirectly related data are necessary for many rare disease applications

>This section will highlight promising approaches for analyzing rare disease data to extract biological insights. 
>We will discuss techniques like transfer learning, representation learning, cascade learning, integrative analysis, and knowledge-graph creation and use that leverage other knowledge and data sources to construct testable hypotheses from rare diseases datasets with limited sample sizes 1–8.

#### Representation learning

Representation learning, also called feature learning, is the process of learning features from raw data, where a feature is an individual variable or property.
An algorithm or approach will construct features as part of training and, in the case of supervised feature learning, use those features to predict labels on input data. 
Using an example from transcriptomics, an unsupervised method such as matrix factorization can be used to extract a low-dimensional representation of the gene-level data, learning features that are a combination of input genes' expression levels [@doi:10.1093/bioinformatics/btq503; @doi:10.1186/s13059-020-02021-3].
Low-dimensional representations trained on compendia of transcriptomic data can then be used as input to supervised machine learning methods [@doi:10.1186/s12859-020-3427-8]. 
Supervised neural networks used in medical imaging studies [@doi:10.1016/j.procs.2016.07.014] (reviewed in [@doi:10.1016/j.zemedi.2018.11.002] and [@doi:10.1098/rsif.2017.0387]), which are trained to predict labels or classes, are also an example of representation learning.
Learned features in the medical imaging domain may be a series of edges that constitute a blood vessel. 
From transcriptomics data, learned features could be coordinated sets of genes involved in a biological process.

In the rare disease domain, Dincer et al. leveraged publicly available acute myeloid leukemia (AML) gene expression data to improve the prediction of _in vitro_ drug responses [@doi:10.1101/278739]. 
The authors trained a variational autoencoder (VAE) on AML data that was unlabeled with regards to drug response, used the learned features to encode a low-dimensional representation of held-out AML data with drug response labels, and used this low-dimensional representation as input to L1-regularized logistic regression.

Representation learning tends to be data-intensive–many samples are required.
Though there were over 6500 publicly available AML samples from many different studies used as part of the training set in Dincer et al. [@doi:10.1101/278739], we expect that in other rare diseases considerably fewer samples will be available for training or, in the case of systemic diseases, studies may be from different tissues.
In the next section, we will introduce methods or approaches that may be more broadly useful in rare diseases; representation learning underlies many of them.

#### Transfer learning

We focus on a series of approaches that are centered on the following concept: to realize the potential of machine learning for biological discovery in rare diseases, we often can not study an individual rare disease alone as samples are limited.
We may instead build on prior knowledge and large volumes of data that do not directly assay our disease of interest, but are closely-enough related to be valuable for discovery.
We can leverage shared features, whether they are biological patterns that are a normal part of development aberrant in a particular disease context or an imaging anomaly present in both rare and common diseases, for advancing our understanding.
We discuss transfer learning or domain adaptation, multitask learning, and few-shot learning approaches below.

* Transfer learning example - mention how representation learning or feature construction can be related/part of this approach
* Multitask learning example
* Few-shot learning example

<!--The section I'm outlining below may need to be moved as the manuscript matures-->

To be successful, machine learning approaches in rare diseases likely require the inclusion _multiple strategies_ discussed throughout.

PLIER and MultiPLIER studies as examples 
* Representation learning from an arbitrary, large set of gene expression data
* Prior knowledge
* Regularization
