### Build upon prior knowledge and indirectly related data {.page_break_before}

Rare diseases often lack large, normalized datasets, limiting our ability to study key attributes of these diseases. 
Thus evaluating genotype-phenotype relationships or repurposing drugs using knowledge graphs can greatly benefit rare disease. 
Knowledge graphs (KGs) integrate related-but-different data types, creating a rich data source (e.g. Monarch Graph Database[@doi:10.1093/nar/gkw1128], hetionet[@doi:10.7554/elife.26726], PheKnowLator[@doi:10.1101/2020.04.30.071407], and the Global Network of Biomedical Relationships[@doi:10.1093/bioinformatics/bty114], Orphanet[@https:http://www.orpha.net]). 
These graphs connect genetic, functional, chemical, clinical, and ontological data to enable the exploration of relationships of data with disease phenotypes through manual review[@doi:10.1093/database/baaa015] or computational methods[@doi:10.1101/727925; @doi:10.1186/s12911-019-0938-1].
KGs may include  links or nodes that are specific to the rare disease of interest (e.g. an FDA approved treatment  would be a specific disease-compound link in the KG ) as well as links that are more generalized (e.g. gene-gene interactions noted in the literature for a different disease). 

Rare disease researchers can leverage the entities and relationships outside of the specific disease-context, including common comorbidities that are more prevalent conditions[@doi:10.1101/727925], for prediction. 
Such approaches have been used in rare disease research in areas such as drug repurposing[@doi:10.1101/727925] and disease classification[@doi:10.1186/s12911-019-0938-1]. 
Identifying KG encoding methods that can provide actionable insights for a specific rare disease application is an active area of research. 
Other approaches that build upon prior knowledge and large volumes of related data include transfer learning, multitask learning, and few-shot learning approaches. 
These approaches leverage shared features, e.g. normal developmental processes that are aberrant in disease, or an imaging anomaly present in rare and common diseases, for advancing our understanding of rare diseases. 

Transfer learning, where a model trained for one task or domain (source domain) is applied to another related task or domain (target domain), can be supervised , or unsupervised . 
Among various types of transfer learning we will mainly focus on feature-representation-transfer. 
Feature-representation-transfer approaches learn representations from the source domain and apply them to a target domain [@doi:10.1109/tkde.2009.191
]. 
For example, low-dimensional representations can be learned from tumor transcriptomic data and transferred to describe patterns associated with genetic alterations in cell line data [@doi:10.1186/s13059-020-02021-3].

Other approaches related to transfer learning, multitask and few-shot learning, are forms of supervised learning that often  rely on deep neural networks. 
In multitask learning classifiers use shared representations to learn multiple related but individual predictions (tasks) simultaneously [@doi:10.1023/a:1007379606734]. Few-shot learning on the other hand generalizes a model trained on related tasks to a new task with limited labeled data (e.g., the detection of a patient with a rare disease from a low number of examples of that rare disease). 
While various approaches and architectures underlie multitask and few-shot learning (see [@https://arxiv.org/abs/1706.05098; @https://arxiv.org/abs/1707.08114v2; @https://arxiv.org/abs/1904.05046v3] for an overview), we will delve into a few selected studies to illustrate potential uses and limitations of these approaches in rare disease.

Examination of the effects of dataset size and task relatedness on multitask learning performance improvements (“multitask effect”) in drug discovery showed that smaller datasets tended to benefit most from multitask learning and the addition of more training data did not guarantee improved performance for multitask models [@https://arxiv.org/abs/1606.08793]. 
Another study demonstrated that performance gains were context-dependent, i.e., multitask neural networks outperformed single-task networks for predicting complex rare phenotypes from EHR data, but not common phenotypes [@https://arxiv.org/abs/1808.03331]. The top-performer in a recent DREAM challenge for predicting drug sensitivity in cancer cell lines, including cell lines from rare cancers, was a multitask learning approach [TODO: CTD-squared Chemogenomic DREAM Challenge citation]. 
From these studies and others, it is clear that multitask learning is a promising approach for rare disease research albeit with some important, context-specific limitations.
In contrast, one-shot or few-shot learning uses prior knowledge to generalize a distance metric learned from input data to compare with a low number of new examples for prediction [@https://arxiv.org/abs/1904.05046v3, @doi:10.1021/acscentsci.6b00367], e.g. a method developed for predicting small molecule activity learned a meaningful distance metric over the properties of various compounds [@doi:10.1021/acscentsci.6b00367]. 
But the authors’ results also suggest that structural similarity among compounds was a requirement for this desired performance boost. 
In a study of rare pathologies in fundus photographs, a few-shot learning approach had a performance advantage over multitask learning, since predicting common conditions simultaneously resulted in a loss of performance for the multitask learner [@doi:10.1016/j.media.2020.101660]. 
Thus transfer, multi-task, and few-shot learning are appealing for the study of rare diseases, conditions, or phenotypes, but their limits and potential utility are still open research questions. 
Nevertheless, selecting an appropriate model for a given task and evaluations that are well-aligned with a research question are crucial for applying these approaches in rare diseases.
