### Utilize prior knowledge and indirectly related data {.page_break_before}

Rare diseases lack large, normalized datasets, limiting our ability to study key attributes of these diseases.
A potentially powerful strategy for evaluating genotype-phenotype relationships or repurposing drugs when large datasets are scarce is to use knowledge graphs.
Knowledge graphs integrate related-but-different data types, creating a rich data source.
Examples of public biomedical knowledge graphs and frameworks that could be useful in rare disease include the Monarch Graph Database[@doi:10.1093/nar/gkw1128], hetionet[@doi:10.7554/eLife.26726], PheKnowLator[@doi:10.1101/2020.04.30.071407], and the Global Network of Biomedical Relationships[@doi:10.1093/bioinformatics/bty114].
These graphs connect information like genetic, functional, chemical, clinical, and ontological data to enable the exploration of relationships of data with disease phenotypes through manual review[@doi:10.1093/database/baaa015] or computational methods[@doi:10.1101/727925; @doi:10.1186/s12911-019-0938-1].

In the academic rare disease space, there a few pioneering examples of ML-based mining of knowledge graphs to repurpose drugs[@doi:10.1101/727925] and classify rare diseases[@doi:10.1186/s12911-019-0938-1] highlight the challenges in using these in rare disease.
For example, these projects rely on a gold standard datasets to validate the performance of the models which are often absent for individual rare diseases.
They also evaluate rare diseases in an unbiased manner, rather than interrogating a specific disease of interest.
Consequently, it is not yet clear how effective these approaches, and knowledge graphs in general, are in studying a specific disease of interest.
More work needs to be done to identify methods that can provide actionable insights for a specific rare disease application. <!-- TODO: Is this a point to note that may need to be combined with statistical techniques described earlier? -->

Beyond the few academic studies, private entities (e.g. healx, Boehringer Ingelheim) are also performing an undisclosed amount of work to create proprietary rare disease knowledge graphs for ML-based drug discovery applications.
The existence of private companies pursuing this idea, as well as the availability of public biomedical knowledge graphs, suggests that this may be a fruitful untapped area of rare disease research in the public arena.
More work needs to be done to assess 1) which graphs and graph features capture the salient information about rare diseases, 2) the utility of ML methods to obtain actionable insights about rare diseases and 3) which problems - like drug discovery, identification of novel rare diseases, or assessment of genotype-phenotype relationships - can be interrogated using ML of knowledge graphs.
However, the existence of interest in this idea in both pharma and public domain suggests that this is an untapped area of rare disease research.

<!-- TODO: Is this header level right? I can't tell if this is supposed to be a subsection or its own section. The strategies discussed here might either be considered bringing both statistical + prior knowledge & data together or another part of using prior knowledge + data. I think I favor the former, though it would require putting a brief transfer and multi-task blurb in the prior/related data section. -->

Other approaches that build upon prior knowledge and large volumes of related data include transfer learning, multitask learning, and few-shot learning approaches.
These approaches leverage shared features e.g. normal developmental processes that are aberrant in disease, imaging anomalies present in rare and common diseases, for advancing our understanding of rare diseases.

Transfer learning is an approach where a model trained for one task or domain (source domain) is applied to another, typically related task or domain (target domain).
Transfer learning can be supervised (one or both of the source and target domains have labels), or unsupervised (both domains are unlabeled).
Among the various types of transfer learning, we will mainly focus on _feature-representation-transfer_.
Feature-representation-transfer approaches learn representations from the source domain and apply them to a target domain [@doi:10.1109/TKDE.2009.191].
For example, low-dimensional representations can be learned from tumor transcriptomic data and transferred to describe patterns associated with genetic alterations in cell line data [@doi:10.1186/s13059-020-02021-3].  
<!--TODO: anything need to be added/revised here once the dimensionality reduction section is rewritten to be more general? -->

Transfer learning can be supervised or unsupervised, but the related approaches multitask and few-shot learning are forms of supervised learning that generally rely on deep neural networks.
Multitask learning is an approach where classifiers use shared representations to learn multiple _related but individual predictions_ (tasks) simultaneously [@doi:10.1023/A:1007379606734].
Few-shot learning on the other hand generalizes a model trained on related tasks to a new task with limited labeled data (e.g., the detection of a patient with a rare disease from a low number of examples of that rare disease).
While the variety of approaches and architectures that underlie multitask and few-shot learning are beyond this scope of this work (see [@arxiv:1706.05098; @arxiv:1707.08114v2] and [@arxiv:1904.05046v3] for an overview), we will delve into a few selected studies to illustrate potential uses and limitations of these approaches in rare disease.

Multitask neural networks (which predict multiple tasks simultaneously) improve model performance over single task models by learning a shared representation, effectively being exposed to more training data than single task models [@doi:10.1023/A:1007379606734; @arxiv:1606.08793].
Examination of the effects of dataset size and task relatedness on multitask learning performance improvements ("multitask effect") in drug discovery showed that smaller datasets tended to benefit most from multitask learning and the addition of more training data did not guarantee improved performance for multitask models [@arxiv:1606.08793].
Another study demonstrated that performance gains were context-dependent, i.e. multitask neural networks outperformed single-task networks for predicting complex rare phenotypes from EHR data, but not common phenotypes [@arxiv:1808.03331].

In contrast, one-shot or few-shot learning uses prior knowledge to generalize a distance metric learned from input data to compare with a low number of new examples for prediction [@arxiv:1904.05046v3; @doi:10.1021/acscentsci.6b00367], e.g. a method developed for predicting small molecule activity learned a meaningful distance metric over the properties of various compounds [@doi:10.1021/acscentsci.6b00367].
But the authors' results also suggest that also suggest limited performance boost or generalizability of the models relative to baseline models in absence of structural similarity among compounds. 
Moreover, few-shot learning approaches outperform multitask learning in contexts where predicting common conditions simultaneously results in a loss of performance [@arxiv:1907.09449v3].  

Thus transfer, multi-task, and few-shot learning are appealing for the study of rare diseases, conditions, or phenotypes, but their limits and potential utility are still open research questions.
Nevertheless, selecting an appropriate model for a given task and evaluations that are well-aligned with a research question are crucial for applying these approaches in rare diseases.
