### Using composite approaches can be a powerful strategy

We have described multiple approaches for maximizing the success of machine learning applications in rare disease research throughout. 
In practice, it is rarely sufficient to use one of these techniques in isolation. 
Below, we highlight two recent works in the rare disease domain that draw on multiple concepts covered in the earlier sections. 
Feature-representation-transfer, which incorporates dimension reduction through representation learning, prior data, and regularization underlie both approaches.

Thousands of acute myeloid leukemia (AML) patient gene expression samples have been collected over time and are publicly available.
Not all of these samples include the most relevant clinical or phenotypic data such as drug response.
However, these publicly available data include an _in vitro_ experiment that examined the response to 160 drugs for 30 AML patient samples, measured using genome-wide arrays which have tens of thousands of features [@doi:10.1038/s41467-017-02465-5].
Training on this drug response dataset alone poses challenges raised throughout –  many features combined with small sample size can result in ML models that are of limited utility and sample size can also be prohibitive when performing representation learning.
Dincer et al. trained a variational autoencoder on over 6500 AML samples (VAE; see [definitions]) [TODO: link between sections?] to reduce the dimensionality of the test set in an approach termed DeepProfile [@https://www.biorxiv.org/content/10.1101/278739v2]. (Figure[@fig:4]a)
The 30 AML test samples with drug response information were encoded using the VAE's low-dimensional representation or _transferred_, reducing the number of features from thousands to eight.
LASSO linear regression models that used encodings as features had better performance than models that used individual gene expression values as features on average.
In addition, the low-dimensional representation learned by the VAE captured more biological pathways than PCA, which may be attributable to the constraints on encodings imposed during the training process [definitions] [TODO: link between sections or cite what is in definitions?].
Similar results were observed for prediction of histopathology in another rare cancer (ovarian) [@https://www.biorxiv.org/content/10.1101/278739v2].

While DeepProfile was centered on training on an individual disease and tissue combination, some rare diseases affect multiple tissues that a researcher may be interested in studying together for the purpose of biological discovery. 
Studying multiple tissues poses significant challenges – features that can be appreciably measured may be tissue-specific even when looking at learned features or representations, a cross-tissue analysis may require an analyst to compare representations from multiple models and models trained on a low number of samples may learn representations that "lump together" multiple biological signals, reducing the interpretability of the results.
To address these challenges, Taroni et al. trained Pathway-Level Information ExtractoR (PLIER) [@doi:10.1038/s41592-019-0456-1] on a large generic collection of human transcriptomic data (recount2 [@doi:10.1038/nbt.3838]).
The authors used the latent variables learned by the model to describe transcriptomic data from the unseen rare diseases antineutrophil cytoplasmic antibody (ANCA)-associated vasculitis (AAV) and medulloblastoma in an approach termed MultiPLIER [@doi:10.1016/j.cels.2019.04.003]. 
(Here "unseen" refers to the fact that these diseases were not in the training set).
PLIER is a matrix factorization approach that takes prior knowledge in the form of gene sets or pathways and gene expression data as input [@doi:10.1038/s41592-019-0456-1]. 
PLIER includes constraints (regularization) such that some latent variables learned by the model will align with input gene sets and ideally latent variables will only be associated with a low number of related gene sets [@doi:10.1038/s41592-019-0456-1], which make it suitable for biological discovery or description of rare disease data. 
MultiPLIER allows us to use one model to describe multiple datasets instead of reconciling output from multiple models, which is highly beneficial when identifying commonalities among disease manifestations or affected tissues is a research goal. (Figure[@fig:4]b)
(This benefit extends to studying multiple cohorts with a different model.)
The inclusion of a large number of samples from diverse biological conditions in the training set results in models with desirable features (e.g., similar pathways are disentangled or separated out in the learned representations).

Taken together, DeepProfile [@doi:10.1101/278739v2] and MultiPLIER [@doi:10.1016/j.cels.2019.04.003] suggest a combination of the techniques discussed throughout this article can be capitalized on for rare disease research. 
In cases where we have few samples from our disease of interest with the required phenotypic labels, we can leverage existing collections of data and knowledge if we select the models with the right attributes. 
The utility of DeepProfile and MultiPLIER stem from the fact that biological processes can be shared between biological contexts and that the methods underlying the approaches can effectively learn about those processes. 
In the natural images field, researchers have demonstrated that the transferability of features depends on relatedness of tasks [@arxiv:1411.1792]. 
The limits of transfer learning for and the concept of relatedness in high-dimensional biomedical data assaying rare diseases are open research questions. 
In the authors' opinion, selecting an appropriate model for a given task and evaluations that are well-aligned with a research goal are crucial for applying these approaches in rare diseases .

![Combining multiple strategies strengthens the performance of ML models in rare disease](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/content/images/figures/pdfs/multiplier-DeepProfile.pdf){#fig:4}
