### Using composite approaches can be a powerful strategy

We have described multiple approaches for maximizing the success of ML applications in rare disease, but it is rarely sufficient to use any of these techniques in isolation. 
Below, we highlight two recent works in the rare disease domain that draw on concepts of feature-representation-transfer, use of prior data, and regularization.

A large public dataset of acute myeloid leukemia (AML) patient samples with no drug response data and a small _in vitro_ experiment with drug response data form the basis of our first example [@doi:10.1038/s41467-017-02465-5].
Training an ML model on the small _in vitro_ dataset alone faced the _curse of dimensionality_ and the dataset size prohibited representation learning.
Dincer et al. trained a variational autoencoder on the large AML patient dataset (VAE; see [definitions]) to learn meaningful representations in an approach termed DeepProfile [@https://www.biorxiv.org/content/10.1101/278739v2] (Figure[@fig:4]a).
The representations or _encodings_ learned by the VAE were then _transferred_ to the small _in vitro_ dataset reducing it's number of features from thousands to eight, and improving the performance of the final LASSO linear regression model.
In addition to improvement in performance, the _encodings_ learned by the VAE captured more biological pathways than PCA, which may be attributable to the constraints on the encodings imposed during the training process [definitions].
Similar results were observed for prediction of histopathology in another rare cancer dataset [@https://www.biorxiv.org/content/10.1101/278739v2].

While DeepProfile was centered on training on an individual disease and tissue combination, some rare diseases affect multiple tissues that a researcher may be interested in studying together for the purpose of biological discovery. 
Studying multiple tissues poses significant challenges and a cross-tissue analysis may require an analyst to compare representations from multiple models.
Models trained on a low number of samples may learn representations that "lump together" multiple biological signals, reducing the interpretability of the results.
To address these challenges, Taroni et al. trained a Pathway-Level Information ExtractoR (PLIER) (a matrix factorization approach that takes prior knowledge in the form of gene sets or pathways) on a large generic collection of human transcriptomic data [@doi:10.1038/s41592-019-0456-1]. 
PLIER used constraints (regularization) that learned _latent variables_ aligned with a small number of input gene sets, making it suitable for biological discovery or description of rare disease data. 
The authors _transferred_ the representations or _latent variables_ learned by the model to describe transcriptomic data from the unseen rare diseases antineutrophil cytoplasmic antibody (ANCA)-associated vasculitis (AAV) and medulloblastoma in an approach termed MultiPLIER [@doi:10.1016/j.cels.2019.04.003]. (Figure[@fig:4]b)
MultiPLIER used one model to describe multiple datasets instead of reconciling output from multiple models, thus making it possible to identify commonalities among disease manifestations or affected tissues. 

Taken together, DeepProfile [@doi:10.1101/278739v2] and MultiPLIER [@doi:10.1016/j.cels.2019.04.003] suggest a combination of the techniques discussed throughout this article can be capitalized on for rare disease research. 
In cases where we have a few samples from our disease of interest with the required phenotypic labels, we can leverage prior knowledge if we select the models with the right attributes. 
DeepProfile and MultiPLIER capitalizes on the fact that biological processes can be shared between biological contexts and that the methods underlying the approaches can effectively learn about those processes. 

![Combining multiple strategies strengthens the performance of ML models in rare disease](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/content/images/figures/pdfs/multiplier-DeepProfile.pdf){#fig:4}
