### Building upon prior knowledge and indirectly related data {.page_break_before}

One strategy to overcome the lack of large normalized datasets in rare disease is to combine a variety of data types and explore rare disease data alongside other knowledge.
By using several data modalities (such as curated pathways, genetic data, or other data types), it may be possible to gain a better understanding of rare diseases (e.g., identifying novel genotype-phenotype relationships or opportunities for drug repurposing).
Knowledge graphs (KGs) which integrate related-but-different data types, provide a rich multimodal data source (e.g., Monarch Graph Database [@doi:10.1093/nar/gkw1128], hetionet [@doi:10.7554/elife.26726], PheKnowLator [@doi:10.1101/2020.04.30.071407], and the Global Network of Biomedical Relationships [@doi:10.1093/bioinformatics/bty114], Orphanet [@http://www.orpha.net]). 
These graphs connect genetic, functional, chemical, clinical, and ontological data so that relationships of data with disease phenotypes can be explored through manual review [@doi:10.1093/database/baaa015] or computational methods [@doi:10.1016/j.jbi.2021.103838, @doi:10.1142/9789811215636_0041; @doi:10.1186/s12911-019-0938-1]. (Figure[@fig:3]a)
KGs may include links (i.e. edges) or nodes that are specific to a rare disease of interest (e.g., an FDA approved treatment would be a specific disease-compound edge in the KG) and more generalized information (e.g., gene-gene interactions noted in the literature for a different disease). (Figure [@fig:4]a)

Rare disease researchers can repurpose general (i.e., not rare disease-specific) biological or chemical KGs to answer rare disease-based research questions [@doi:10.1142/9789811215636_0041] (e.g. Box 1b).
One tactic to sift through the large amounts of data encoded in KGs is to calculate the distances between nodes of interest (e.g., diseases and drugs for Box 1b [@doi:10.1142/9789811215636_0041]); often done by determining the "embeddings" (linear representations of the position and connections of a particular point in the graph) for nodes in the KG, and calculating the similarity between these embeddings.
Effective methods to calculate node embeddings that can generate actionable insights for rare diseases is an active area of research [@doi:10.1142/9789811215636_0041].

Another application of KGs is to augment a dataset [@doi:10.1186/s12911-019-0752-9, doi:10.1186/s12911-019-0938-1].
Li et. al.[@doi:10.1186/s12911-019-0938-1] used a KG to identify linked terms in a medical corpus from a large number of patients, some with rare disease diagnoses.
They augmented their text dataset by map related clinical terms together - e.g., mapping "cancer" and "malignancy" in different patients to the same clinical concept.
With this improved dataset, they trained and tested a variety of text classification algorithms to identify rare disease patients within their corpus. (Figure [@fig:4]b, Box 1a)

Rare disease researchers have also integrated multiple KGs and applied neural network-based algorithms optimized for graph data, such as a graph convolutional neural network.
Rao and colleagues [@doi:10.1186/s12920-018-0372-8] describe the construction of a KG using phenotype information (Human Phenotype Ontology) and rare disease information (Orphanet) and curated gene interaction/pathway data (Lit-BM-13, WikiPathways) [@pmc:PMC7778952; @doi:10.1016/j.cell.2014.10.050; @doi:10.1093/nar/gkaa1024]. 
They then trained a spectral graph convolution neural network on this KG to identify and rank potentially causal genes for the rare diseases from Orphanet, and were able to use this model to accurately predict causal genes for a ground truth dataset of rare diseases with known causal genes.
While several groups have used KGs to study rare diseases, we expect that better multi-modal datasets and ML methods to analyze KGs will make them a more popular and important tool in rare disease. 

Another approach that builds on prior knowledge and large volumes of related data is transfer learning.
Transfer learning leverages shared features, e.g., normal developmental processes that are aberrant in disease or an imaging anomaly present in both rare and common diseases, to advance our understanding of rare diseases. 
Transfer learning, where a model trained for one task or domain (source domain) is applied to another related task or domain (target domain), can be supervised or unsupervised.
Among various types of transfer learning, feature-representation-transfer approaches learn representations from the source domain and apply them to a target domain [@doi:10.1109/tkde.2009.191](Figure [@fig:5]a-c).
That is, representation learning, as discussed earlier, does not need to be applied only to describe the dataset on which the algorithm was trained â€“ it can also be used to elucidate signals in sufficiently similar data (Figure [@fig:5]c) and may offer an improvement in descriptive capability over models trained only on small rare disease datasets (Fig [@fig:5]c).
For instance, low-dimensional representations can be learned from tumor transcriptomic data and transferred to describe patterns in genetic alterations in cell lines [@doi:10.1186/s13059-020-02021-3](Figure [@fig:5]c).
In the next section, we summarize specific instances of applying transfer learning, along with other techniques, to the study of rare diseases.
